nsq/reader.py

```diff
@@ -660,6 +670,17 @@ def _redistribute_rdy_state(self):
         if self.need_rdy_redistributed:
             self.need_rdy_redistributed = False
 
+            if self.total_rdy > self.max_in_flight:
```

Comment:
Should this block instead just set RDY 0 for all connections and let the rest of the logic below identify conns to receive non-zero ready?

Reply:
With the current redistribution logic I think that change could make it unsafe to call `set_max_in_flight()` with messages in flight. Right now redistribution avoids giving positive RDY to a connection with a message in flight, so something like the following could happen:

1. Let max-in-flight be  > `len(conns)`.
2. With a message in flight on conn `c`, reduce m-i-f but  keep it >= `len(conns)`.
3. `_redistribute_rdy_state()` sets RDY 0 on all connections, including `c`.
4. Because `c` has something in flight, this redistribution round leaves `c` at RDY 0.
5. Subsequent redistribution rounds see m-i-f >= `len(conns)` and so don't redistribute RDY, leaving `c` starved.

We could make the simplification you suggest if we allow redistribution to give RDY to a conn with something in flight:
```
--- a/nsq/reader.py
+++ b/nsq/reader.py
@@ -720,7 +720,7 @@ class Reader(Client):
             # We also don't attempt to avoid the connections who previously might have had RDY 1
             # because it would be overly complicated and not actually worth it (ie. given enough
             # redistribution rounds it doesn't matter).
-            possible_conns = [c for c in conns if not (c.in_flight or c.rdy)]
+            possible_conns = [c for c in conns if not c.rdy]
             while possible_conns and available_rdy:
                 available_rdy -= 1
                 conn = possible_conns.pop(random.randrange(len(possible_conns)))
```
which (I think) is fine from a correctness standpoint and will at worst increase the variance in how long it takes to service all connections equally.